tachyon/math/base/rings.h

```diff
@@ -25,26 +26,51 @@ class Ring : public AdditiveGroup<F>, public MultiplicativeSemigroup<F> {
   // This is taken and modified from
   // https://github.com/arkworks-rs/algebra/blob/5dfeedf560da6937a5de0a2163b7958bd32cd551/ff/src/fields/mod.rs#L298C1-L305
   // Sum of products: a₁ * b₁ + a₂ * b₂ + ... + aₙ * bₙ
-  template <
-      typename InputIterator,
-      std::enable_if_t<std::is_same_v<F, base::iter_value_t<InputIterator>>>* =
-          nullptr>
-  constexpr static F SumOfProducts(InputIterator a_first, InputIterator a_last,
-                                   InputIterator b_first,
-                                   InputIterator b_last) {
-    F sum = F::Zero();
-    while (a_first != a_last) {
-      sum += (*a_first * *b_first);
-      ++a_first;
-      ++b_first;
-    }
-    return sum;
+  // TODO(chokobole): If I call |SumOfProducts()| instead of
+  // |SumOfProductsSerial| for all call sites, it gets stuck when doing
+  // unittests. I think we need a some general threshold to check whether it is
+  // good to doing parallelization.
+  template <typename Container>
+  constexpr static F SumOfProducts(const Container& a, const Container& b) {
+    size_t size = std::size(a);
+    CHECK_EQ(size, std::size(b));
+    CHECK_NE(size, size_t{0});
```

Comment:
We have decided not to cast to `size_t` when the actual value is not being used.

Reply:
```c++
int main() {                                                                                                  
  int a = 3; 
  unsigned int b = 4; 
  return a != b;
}
```

It gives an error like following!

```shell
> g++ -Wall -Werror a.cc
a.cc: In function ‘int main()’:
a.cc:4:12: error: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Werror=sign-compare]
    4 |   return a != b;
      |          ~~^~~~
cc1plus: all warnings being treated as errors
```